### ISSUE TITLE
Assertion `false` failed: No Adapter From Version $19 for Constant

### ISSUE THREAD

**Issue by RabJon** [onnx/onnx#5810]
# Bug Report

### Is the issue related to model conversion?

Yes, it is. However, the issue does not appear when converting my TensorFlow model to Onnx, but only later when I run `onnx.version_converter.convert_version()` to convert between Onnx Opset versions. Therefore, I decided to post the issue in this repo.

### Describe the bug
My goal is to convert a TensorFlow model to Onnx with a "current" version of Onnx. I managed to convert my models using Onnx version 1.12.0, but not with versions 1.15.0 and 1.14.1.

When I tried the newest version I came across the bug reported in this issue: https://github.com/onnx/tensorflow-onnx/issues/2262. As suggested in the comments I downgraded Onnx to 1.14.1, which is not a fix I like, but I tried it anyway. Afterwards, this first issue was solved, but I ran into another one:

> RuntimeError: /github/workspace/onnx/version_converter/BaseConverter.h:68: adapter_lookup: Assertion `false` failed: No Adapter From Version $19 for Constant

In my particular case I tried to use the function `onnx.version_converter.convert_version()` to convert to opset version 16, but I also don't mind if it is a version >=16. Therefore, I also tried the target opset versions 18 and 19. However, it appears that for versions below 19 I will always run into the issue reported here and if I use version 19 I run into the following exception: 

> ValueError: make_sure failure: Opset 19 is not supported yet. Please use a lower opset

### System information

- OS Platform and Distribution (*e.g. Linux Ubuntu 20.04*): Linux Ubuntu 22.04.3 LTS
- ONNX version (*e.g. 1.13*): 1.14.1
- Python version: 3.11.5
- GCC/Compiler version (if compiling from source): /
- CMake version: /
- Protobuf version: 3.20.3
- Visual Studio version (if applicable): /

### Reproduction instructions

```
import onnx
from onnx import version_converter

m_path = "<PATH TO SOME ONNX MODEL>"
onnx_model = onnx.load(m_path)
converted_model = version_converter.convert_version(onnx_model , 16)

```

### Expected behavior

The conversion should work as it did for earlier versions of Onnx.

**Comment by justinchuby** [comment]
Downgrade is less well supported at this point. You may add a line in https://github.com/onnx/onnx/blob/67c456ba4747412afb44158a1a889c0fc3349641/onnx/version_converter/convert.h#L566 

```
registerAdapter(std::make_unique<CompatibleAdapter>("Constant", OpSetID(19), OpSetID(18)));
```

that does the reverse

**Comment by RabJon** [comment]
@justinchuby thank you for the fast reply!

I would understand that a downgrade is less supported, if the exception itself didn't tell me to downgrade.
As mentioned I tried it with Opset version 19, which, and correct me if I am wrong, is the most current Opset version and would also be fine for me (like any version >=16). 
However, when using version 19 the exception tells me to use a lower version and now you are telling me that I need to touch the underlying C++ code to do that. I am sorry, but this doesn't seem like a long-term solution to me.

**Comment by justinchuby** [comment]
Is the error message that asked you to downgrade from tf2onnx?

**Comment by RabJon** [comment]
> Is the error message that asked you to downgrade from tf2onnx?

Yes, it is. I just saw that right now. My mistake. So, if I am using opset version 19. The following code raises the following exception: 

```python
tf2onnx.convert.from_keras(model, output_path=m_path, opset=onnx_opset)
```

```
ValueError: make_sure failure: Opset 19 is not supported yet. Please use a lower opset"
```

However, when using a lower Opset version than 19 I get the error from the version_converter.convert_version() function that I reported in my issue.

**Comment by justinchuby** [comment]
@fatcat-z could you help? Thanks!

**Comment by fatcat-z** [comment]
@RabJon , sorry for the inconvenience. The issue [2262](https://github.com/onnx/tensorflow-onnx/issues/2262) you mentioned is going to be fixed by [this PR](https://github.com/onnx/tensorflow-onnx/pull/2279). With a new release of TF2ONNX later, you can use the latest version of ONNX to work.

Currently, I see you've already converted your TF model to an existing ONNX file. So please upgrade your ONNX to latest version 1.15.0, and then call your code "converted_model = version_converter.convert_version(onnx_model , 19)", it should convert the opset version to 19.

I tried this workaround for one ONNX file and it works. Please feel free to let me know if it doesn't work for you.

**Comment by RabJon** [comment]
Thanks for your help @fatcat-z. So, if I understand you right, you suggest to convert the TensorFlow model to an Onnx model using Onnx version 1.14.1 and then to convert the generated Onnx model to Opset 19 with Onnx version 1.15.0 ?

Well, what I tried again now is to use only Onnx version 1.15.0 and set opset versions 19 and 20 as targets, but both failed with the issue https://github.com/onnx/tensorflow-onnx/issues/2262 you mentioned.

My original goal was to update all my dependencies to the newest versions, because I was slowly getting out-of-date, but I still have my old, working environment with which I can convert models up to opset 16. So, I am fine for the moment and will not spend time on any further tests, if the problems I face right now will be solved by your PR. Can you tell me approx. when the release including your fix will be published?

**Comment by fatcat-z** [comment]
> Thanks for your help @fatcat-z. So, if I understand you right, you suggest to convert the TensorFlow model to an Onnx model using Onnx version 1.14.1 and then to convert the generated Onnx model to Opset 19 with Onnx version 1.15.0 ?
> 
> Well, what I tried again now is to use only Onnx version 1.15.0 and set opset versions 19 and 20 as targets, but both failed with the issue [onnx/tensorflow-onnx#2262](https://github.com/onnx/tensorflow-onnx/issues/2262) you mentioned.
> 
> My original goal was to update all my dependencies to the newest versions, because I was slowly getting out-of-date, but I still have my old, working environment with which I can convert models up to opset 16. So, I am fine for the moment and will not spend time on any further tests, if the problems I face right now will be solved by your PR. Can you tell me approx. when the release including your fix will be published?

"convert the TensorFlow model to an Onnx model using Onnx version 1.14.1 and then to convert the generated Onnx model to Opset 19 with Onnx version 1.15.0 ?" - Yes, you are right.

My plan is to deliver this release before Dec 29, and will update here if the latest version of tf2onnx has been released.

**Comment by RabJon** [comment]
> My plan is to deliver this release before Dec 29, and will update here if the latest version of tf2onnx has been released.

That sounds great! Thank you for your help @fatcat-z and @justinchuby

**Comment by fatcat-z** [comment]
> > My plan is to deliver this release before Dec 29, and will update here if the latest version of tf2onnx has been released.
> 
> That sounds great! Thank you for your help @fatcat-z and @justinchuby

@RabJon ,

[Here](https://github.com/onnx/tensorflow-onnx/releases/tag/v1.16.0) is the latest version of tf2onnx which supports ONNX 1.15.0. You can uninstall your tf2onnx and install the latest one to have a try.

**Comment by RabJon** [comment]
@fatcat-z thank you for your effort!

I updated my packages now to ONNX 1.15.0 and tf2onnx 1.16.0. However, if I try to use opset version 19/20 I still run into this exception: `ValueError: make_sure failure: Opset 19/20 is not supported yet. Please use a lower opset` and if I use a lower opset I run into the exception that I originally reported in this issue.

So, it seems that the issue https://github.com/onnx/tensorflow-onnx/issues/2262 was solved, but the problem with the function `onnx.version_converter.convert_version()` still exists.

**Comment by RabJon** [comment]
It's fine for me if versions 19/20 are not supported yet, but if this is the case, then IMO I should be able to convert back to version 18 or below using `onnx.version_converter.convert_version()`. Or do I misunderstand something?

**Comment by RabJon** [comment]
Are there any updates regarding this issue? I upgraded to the newest tf2onnx version (from 1.16.0 to 1.16.1), but the issue persists.

**Comment by LOMO7711** [comment]
I solved this problem for myself by checking if my model actually needs operators that were introduced or changed after the release of the version I want.
If this is not the case, you can simply hardcode an older version of the operator in your onnx file to the version you need.
